### Importing Libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
psdata=pd.read_csv('DATA/Google-Playstore.csv')

### Data at a first glance :

In [3]:
psdata.head(5)

App Name  \
0                                            Gakondo   
1                                Ampere Battery Info   
2                                             Vibook   
3  Smart City Trichy Public Service Vehicles 17UC...   
4                                            GROW.me   

                       App Id       Category  Rating  Rating Count Installs  \
0         com.ishakwe.gakondo      Adventure     0.0           0.0      10+   
1  com.webserveis.batteryinfo          Tools     4.4          64.0   5,000+   
2        com.doantiepvien.crm   Productivity     0.0           0.0      50+   
3     cst.stJoseph.ug17ucs548  Communication     5.0           5.0      10+   
4        com.horodyski.grower          Tools     0.0           0.0     100+   

   Minimum Installs  Maximum Installs  Free  Price  ...  \
0              10.0                15  True    0.0  ...   
1            5000.0              7662  True    0.0  ...   
2              50.0                58  True    0.0  ...   
3              10.0                19  True    0.0  ...   
4             100.0               478  True    0.0  ...   

                  Developer Website              Developer Email  \
0       https://beniyizibyose.tk/#/       jean21101999@gmail.com   
1   https://webserveis.netlify.app/         webserveis@gmail.com   
2                               NaN          vnacrewit@gmail.com   
3  http://www.climatesmarttech.com/  climatesmarttech2@gmail.com   
4       http://www.horodyski.com.pl    rmilekhorodyski@gmail.com   

       Released  Last Updated Content Rating  \
0  Feb 26, 2020  Feb 26, 2020       Everyone   
1  May 21, 2020  May 06, 2021       Everyone   
2   Aug 9, 2019  Aug 19, 2019       Everyone   
3  Sep 10, 2018  Oct 13, 2018       Everyone   
4  Feb 21, 2020  Nov 12, 2018       Everyone   

                                      Privacy Policy Ad Supported  \
0                 https://beniyizibyose.tk/projects/        False   
1  https://dev4phones.wordpress.com/licencia-de-uso/         True   
2  https://www.vietnamairlines.com/vn/en/terms-an...        False   
3                                                NaN         True   
4                        http://www.horodyski.com.pl        False   

  In App Purchases Editors Choice         Scraped Time  
0            False          False  2021-06-15 20:19:35  
1            False          False  2021-06-15 20:19:35  
2            False          False  2021-06-15 20:19:35  
3            False          False  2021-06-15 20:19:35  
4            False          False  2021-06-15 20:19:35  

[5 rows x 24 columns]

In [4]:
psdata.info(show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2312944 entries, 0 to 2312943
Data columns (total 24 columns):
 #   Column             Non-Null Count    Dtype  
---  ------             --------------    -----  
 0   App Name           2312939 non-null  object 
 1   App Id             2312944 non-null  object 
 2   Category           2312944 non-null  object 
 3   Rating             2290061 non-null  float64
 4   Rating Count       2290061 non-null  float64
 5   Installs           2312837 non-null  object 
 6   Minimum Installs   2312837 non-null  float64
 7   Maximum Installs   2312944 non-null  int64  
 8   Free               2312944 non-null  bool   
 9   Price              2312944 non-null  float64
 10  Currency           2312809 non-null  object 
 11  Size               2312748 non-null  object 
 12  Minimum Android    2306414 non-null  object 
 13  Developer Id       2312911 non-null  object 
 14  Developer Website  1552109 non-null  object 
 15  Developer Email    2312913 non-n

### Data Cleaning

#### Identifing Null values

In [5]:
psdata.isnull().sum()*(100/2312944)

App Name              0.000216
App Id                0.000000
Category              0.000000
Rating                0.989345
Rating Count          0.989345
Installs              0.004626
Minimum Installs      0.004626
Maximum Installs      0.000000
Free                  0.000000
Price                 0.000000
Currency              0.005837
Size                  0.008474
Minimum Android       0.282324
Developer Id          0.001427
Developer Website    32.894657
Developer Email       0.001340
Released              3.071972
Last Updated          0.000000
Content Rating        0.000000
Privacy Policy       18.199879
Ad Supported          0.000000
In App Purchases      0.000000
Editors Choice        0.000000
Scraped Time          0.000000
dtype: float64

- #### Except Developer Website (32.894657 % nulls) and Privacy Policy ( 18.199879 % nulls), all other null rows shall be dropped

In [6]:
l=[ _ for _ in psdata.columns if _ not in ['Privacy Policy','Developer Website']]
psdata[l].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2312944 entries, 0 to 2312943
Data columns (total 22 columns):
 #   Column            Dtype  
---  ------            -----  
 0   App Name          object 
 1   App Id            object 
 2   Category          object 
 3   Rating            float64
 4   Rating Count      float64
 5   Installs          object 
 6   Minimum Installs  float64
 7   Maximum Installs  int64  
 8   Free              bool   
 9   Price             float64
 10  Currency          object 
 11  Size              object 
 12  Minimum Android   object 
 13  Developer Id      object 
 14  Developer Email   object 
 15  Released          object 
 16  Last Updated      object 
 17  Content Rating    object 
 18  Ad Supported      bool   
 19  In App Purchases  bool   
 20  Editors Choice    bool   
 21  Scraped Time      object 
dtypes: bool(4), float64(4), int64(1), object(13)
memory usage: 326.5+ MB


In [7]:
psdata[l].isnull().any(axis=1).value_counts()

False    2235250
True       77694
Name: count, dtype: int64

### Percentage of rows to be removed :

In [8]:
100* psdata[l].isnull().any(axis=1).value_counts()[1]/len(psdata[l])

3.359095594186457

In [9]:
psdata[l].isnull().any(axis=1)

0          False
1          False
2          False
3          False
4          False
           ...  
2312939     True
2312940    False
2312941    False
2312942    False
2312943    False
Length: 2312944, dtype: bool

In [10]:
psdata_cleaned = psdata[~psdata[l].isnull().any(axis=1)]
psdata_cleaned.head(5)

App Name  \
0                                            Gakondo   
1                                Ampere Battery Info   
2                                             Vibook   
3  Smart City Trichy Public Service Vehicles 17UC...   
4                                            GROW.me   

                       App Id       Category  Rating  Rating Count Installs  \
0         com.ishakwe.gakondo      Adventure     0.0           0.0      10+   
1  com.webserveis.batteryinfo          Tools     4.4          64.0   5,000+   
2        com.doantiepvien.crm   Productivity     0.0           0.0      50+   
3     cst.stJoseph.ug17ucs548  Communication     5.0           5.0      10+   
4        com.horodyski.grower          Tools     0.0           0.0     100+   

   Minimum Installs  Maximum Installs  Free  Price  ...  \
0              10.0                15  True    0.0  ...   
1            5000.0              7662  True    0.0  ...   
2              50.0                58  True    0.0  ...   
3              10.0                19  True    0.0  ...   
4             100.0               478  True    0.0  ...   

                  Developer Website              Developer Email  \
0       https://beniyizibyose.tk/#/       jean21101999@gmail.com   
1   https://webserveis.netlify.app/         webserveis@gmail.com   
2                               NaN          vnacrewit@gmail.com   
3  http://www.climatesmarttech.com/  climatesmarttech2@gmail.com   
4       http://www.horodyski.com.pl    rmilekhorodyski@gmail.com   

       Released  Last Updated Content Rating  \
0  Feb 26, 2020  Feb 26, 2020       Everyone   
1  May 21, 2020  May 06, 2021       Everyone   
2   Aug 9, 2019  Aug 19, 2019       Everyone   
3  Sep 10, 2018  Oct 13, 2018       Everyone   
4  Feb 21, 2020  Nov 12, 2018       Everyone   

                                      Privacy Policy Ad Supported  \
0                 https://beniyizibyose.tk/projects/        False   
1  https://dev4phones.wordpress.com/licencia-de-uso/         True   
2  https://www.vietnamairlines.com/vn/en/terms-an...        False   
3                                                NaN         True   
4                        http://www.horodyski.com.pl        False   

  In App Purchases Editors Choice         Scraped Time  
0            False          False  2021-06-15 20:19:35  
1            False          False  2021-06-15 20:19:35  
2            False          False  2021-06-15 20:19:35  
3            False          False  2021-06-15 20:19:35  
4            False          False  2021-06-15 20:19:35  

[5 rows x 24 columns]

In [11]:
psdata_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2235250 entries, 0 to 2312943
Data columns (total 24 columns):
 #   Column             Dtype  
---  ------             -----  
 0   App Name           object 
 1   App Id             object 
 2   Category           object 
 3   Rating             float64
 4   Rating Count       float64
 5   Installs           object 
 6   Minimum Installs   float64
 7   Maximum Installs   int64  
 8   Free               bool   
 9   Price              float64
 10  Currency           object 
 11  Size               object 
 12  Minimum Android    object 
 13  Developer Id       object 
 14  Developer Website  object 
 15  Developer Email    object 
 16  Released           object 
 17  Last Updated       object 
 18  Content Rating     object 
 19  Privacy Policy     object 
 20  Ad Supported       bool   
 21  In App Purchases   bool   
 22  Editors Choice     bool   
 23  Scraped Time       object 
dtypes: bool(4), float64(4), int64(1), object(15)
memory usa

### Checking null values :

In [12]:
psdata_cleaned.isna().sum()

App Name                  0
App Id                    0
Category                  0
Rating                    0
Rating Count              0
Installs                  0
Minimum Installs          0
Maximum Installs          0
Free                      0
Price                     0
Currency                  0
Size                      0
Minimum Android           0
Developer Id              0
Developer Website    740562
Developer Email           0
Released                  0
Last Updated              0
Content Rating            0
Privacy Policy       405439
Ad Supported              0
In App Purchases          0
Editors Choice            0
Scraped Time              0
dtype: int64

- #### As expected, every row has been cleaned ignoring columns 'Developer Website' and 'Privacy Policy' 

### checking for duplicates :

In [13]:
psdata_cleaned.duplicated().sum()

0

#### no duplicates found

## Data formatting : 

### Resetting index and dropping old index

In [14]:
psdata_cleaned.reset_index(inplace=True,drop=True)

### Formatting dates

In [15]:
psdata_cleaned['Released'] = pd.to_datetime(psdata_cleaned['Released'])

C:\Users\Allo\AppData\Local\Temp\ipykernel_3160\3226308707.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  psdata_cleaned['Released'] = pd.to_datetime(psdata_cleaned['Released'])


In [16]:
psdata_cleaned['Released']

0         2020-02-26
1         2020-05-21
2         2019-08-09
3         2018-09-10
4         2020-02-21
             ...    
2235245   2018-05-22
2235246   2018-01-17
2235247   2018-08-19
2235248   2016-08-01
2235249   2019-08-09
Name: Released, Length: 2235250, dtype: datetime64[ns]

In [17]:
psdata_cleaned['Last Updated']=pd.to_datetime(psdata_cleaned['Last Updated'])

C:\Users\Allo\AppData\Local\Temp\ipykernel_3160\903406289.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  psdata_cleaned['Last Updated']=pd.to_datetime(psdata_cleaned['Last Updated'])


In [24]:

psdata_cleaned['Scraped Time'] = pd.to_datetime(psdata_cleaned['Scraped Time'])

C:\Users\Allo\AppData\Local\Temp\ipykernel_3160\758096163.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  psdata_cleaned['Scraped Time'] = pd.to_datetime(psdata_cleaned['Scraped Time'])


In [25]:
psdata_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2235250 entries, 0 to 2235249
Data columns (total 24 columns):
 #   Column             Dtype         
---  ------             -----         
 0   App Name           object        
 1   App Id             object        
 2   Category           object        
 3   Rating             float64       
 4   Rating Count       float64       
 5   Installs           object        
 6   Minimum Installs   float64       
 7   Maximum Installs   int64         
 8   Free               bool          
 9   Price              float64       
 10  Currency           object        
 11  Size               object        
 12  Minimum Android    object        
 13  Developer Id       object        
 14  Developer Website  object        
 15  Developer Email    object        
 16  Released           datetime64[ns]
 17  Last Updated       datetime64[ns]
 18  Content Rating     object        
 19  Privacy Policy     object        
 20  Ad Supported       bool 

### Adding new Columns

In [26]:
psdata_cleaned['period_after_release'] = psdata_cleaned['Scraped Time'] - psdata_cleaned['Released']

psdata_cleaned['period_after_release']

C:\Users\Allo\AppData\Local\Temp\ipykernel_3160\2005866426.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  psdata_cleaned['period_after_release'] = psdata_cleaned['Scraped Time'] - psdata_cleaned['Released']


0          475 days
1          390 days
2          676 days
3         1009 days
4          480 days
             ...   
2235245   1121 days
2235246   1246 days
2235247   1032 days
2235248   1780 days
2235249    677 days
Name: period_after_release, Length: 2235250, dtype: timedelta64[ns]

In [27]:
psdata_cleaned.period_after_release.describe()

count                         2235250
mean     1021 days 17:13:01.993602544
std       773 days 18:04:41.735487880
min                   0 days 00:00:00
25%                 427 days 00:00:00
50%                 819 days 00:00:00
75%                1446 days 00:00:00
max                4157 days 00:00:00
Name: period_after_release, dtype: object

In [28]:
psdata_cleaned['period_after_update'] = psdata_cleaned['Scraped Time'] - psdata_cleaned['Last Updated']
psdata_cleaned['period_after_update']

C:\Users\Allo\AppData\Local\Temp\ipykernel_3160\1929766014.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  psdata_cleaned['period_after_update'] = psdata_cleaned['Scraped Time'] - psdata_cleaned['Last Updated']


0          475 days
1           40 days
2          666 days
3          976 days
4          946 days
             ...   
2235245    915 days
2235246   1230 days
2235247   1032 days
2235248     42 days
2235249    667 days
Name: period_after_update, Length: 2235250, dtype: timedelta64[ns]

In [29]:
psdata_cleaned.period_after_update.describe()

count                        2235250
mean     560 days 03:47:10.578414080
std      556 days 04:44:17.939184696
min                  0 days 00:00:00
25%                132 days 00:00:00
50%                398 days 00:00:00
75%                793 days 00:00:00
max               4510 days 00:00:00
Name: period_after_update, dtype: object

### Formatting the 'Installs' Column :

In [31]:
psdata_cleaned.Installs = psdata_cleaned.Installs.str.replace(',','').str.strip('+').astype('int64')


C:\Users\Allo\AppData\Local\Temp\ipykernel_3160\3315530499.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  psdata_cleaned.Installs = psdata_cleaned.Installs.str.replace(',','').str.strip('+').astype('int64')


In [61]:
psdata_cleaned.Installs.describe()

count    2.235250e+06
mean     1.760321e+05
std      1.420412e+07
min      0.000000e+00
25%      5.000000e+01
50%      5.000000e+02
75%      5.000000e+03
max      1.000000e+10
Name: Installs, dtype: float64

### Filtering data with repect to currency


In [59]:
100 * psdata_cleaned.Currency.value_counts()/len(psdata_cleaned)

Currency
USD    100.0
Name: count, dtype: float64

- #### 100% of the cleaned data has'USD' as currency

## Exporting data for Analysis Phase

In [1]:
#psdata_cleaned.to_csv('psdata_cleaned.csv', header=True, index=False)